src/utils/include/utils/mpi/gather_buffer.hpp

```diff
@@ -104,6 +117,13 @@ void gather_buffer(std::vector<T, Allocator> &buffer,
     /* Resize the buffer */
     buffer.resize(tot_size);
 
+    /* Move the original data to its new location */
```

Comment:
btw: why is there so much code duplication? can we get rid of it?

Reply:
The `std::vector` version takes care of memory allocation. The pointer version doesn't and assumes the caller took care of calling `realloc()`. Once we're done removing all the manual memory allocation in the core, the pointer version can be removed, and all the code duplication will be gone.